Update Vm#find_all_by_mac_address_and_hostname_and_ipaddress

kbrock opened the pull request:
Instead of going to `Hardware` and mapping to `Vm`, just go to `Vm`.
Instead of bringing back a bunch of `Nics` (guest_devices) and `Networks`, just us `joins`.

Performance characteristics:

- `Vm` (vs `VmOrTemplate`) tacked on a bunch of `type` values in the `WHERE` clause.
- The `SELECT` only brings back `vms.*` rather than `hardwares.*`, `vms.*`, and at least one more: `devices.*` or `networks.*`.
- Only instantiates `Vm` on the ruby side vs the extra objects

It is still performing 1 query, but 2 fewer extra tables.
This is now a `scope` rather than an `Array`.

pulled out of https://github.com/ManageIQ/manageiq/pull/23607

Fryguy commented on spec/models/vm_spec.rb:
```diff
@@ -49,39 +49,48 @@
-      expect(described_class.find_all_by_mac_address_and_hostname_and_ipaddress(nil, nil, ipaddress))
-        .to eql([@vm1])
+      expect do
+        expect(described_class.find_all_by_mac_address_and_hostname_and_ipaddress(nil, hostname, nil)).to eq([vm1])
```
I think this should be passing both hostname and Ipaddress as per the rest title right?


```suggestion
        expect(described_class.find_all_by_mac_address_and_hostname_and_ipaddress(nil, hostname, ipaddress)).to eq([vm1])
```

Fryguy commented on app/models/vm.rb:
```diff
@@ -45,35 +45,14 @@ def enforce_single_service_parent?
-            .references(references.uniq)
-            .where(conds)
-            .map(&:vm_or_template).select { |vm| vm.kind_of?(Vm) }
+    return Vm.none if mac_address.blank? && hostname.blank? && ipaddress.blank?
+    scope = Vm
```
Minor rubocop, but there should be a blank line after a guard clause


```suggestion
    return Vm.none if mac_address.blank? && hostname.blank? && ipaddress.blank?

    scope = Vm
```

kbrock commented:
update:
- rubocop space
- Don't auto load the scope

update
- added tests with mac address and hostname (from a different vm)
- moved method from Vm to EvmServer

Comments: I think the mac and hostname should have returned 2 servers, but I ensured the test did not change behavior. So by this, I wonder if `set_local_server_vm` resets the vm properly.

kbrock commented:
update:

- extra spec (different ip and hostname)
- better title for the multi address/ip test

kbrock commented:
So, I'm concerned that `set_local_server_vm` does not work as intended.

Currently if the hostname and ip address point to 2 different machines, then nothing comes back.

I guess having the same `hostname` on multiple `Vm` records would allow multiple `Vm` records to be returned.

@agrare @Fryguy could either of you verify that this is working as intended?

kbrock commented:
thanks.

I was thinking it would be used to migrate from one vm to another. but sounds like these features are mostly focused on tiebreaker scenarios.
